huaweicloud/services/obs/resource_huaweicloud_obs_bucket_replication.go

```diff
@@ -0,0 +1,299 @@
+
+func flattenDestinationBucket(output *obs.GetBucketReplicationOutput) string {
+	rules := output.ReplicationRules
+	if len(rules) == 0 {
+		return ""
+	}
+	return rules[0].DestinationBucket
+}
+
+func convertDestinationStatus(status obs.RuleStatusType) bool {
+	if status == obs.RuleStatusEnabled {
+		return true
+	}
+	return false
+}
+
+func convertDestinationHistoryObjectReplication(historyObjectReplication obs.EnabledType) bool {
+	if historyObjectReplication == obs.Enabled {
+		return true
+	}
+	return false
+}
+
+func flattenDestinationRules(output *obs.GetBucketReplicationOutput) []map[string]interface{} {
+	rules := output.ReplicationRules
+	if len(rules) == 0 {
+		return nil
+	}
+	result := make([]map[string]interface{}, len(rules))
+	for i, rule := range rules {
+		result[i] = map[string]interface{}{
+			"prefix":          rule.Prefix,
+			"storage_class":   rule.StorageClass,
+			"enabled":         convertDestinationStatus(rule.Status),
+			"history_enabled": convertDestinationHistoryObjectReplication(rule.HistoricalObjectReplication),
+			"id":              rule.ID,
+		}
+	}
+	return result
+}
+
+func resourceObsBucketReplicationRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	obsClient, err := cfg.ObjectStorageClientWithSignature(region)
+	if err != nil {
+		return diag.Errorf("error creating OBS Client: %s", err)
+	}
+
+	output, err := obsClient.GetBucketReplication(d.Id())
+	if err != nil {
+		if obsError, ok := err.(obs.ObsError); ok && obsError.Code == "ReplicationConfigurationNotFoundError" {
+			// The bucket does not have replication configurations
+			d.SetId("")
+			mErr := multierror.Append(nil,
+				d.Set("bucket", ""),
+				d.Set("destination_bucket", ""),
+				d.Set("agency", ""),
+				d.Set("rule", nil),
+			)
+			if err := mErr.ErrorOrNil(); err != nil {
+				return diag.Errorf("error setting OBS bucket replication fields: %s", err)
+			}
+			return nil
+		}
+		return diag.FromErr(getObsError("Error retrieving OBS bucket replication", d.Id(), err))
+	}
+
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("bucket", d.Id()),
+		d.Set("destination_bucket", flattenDestinationBucket(output)),
+		d.Set("agency", output.Agency),
+		d.Set("rule", flattenDestinationRules(output)),
+	)
+	if err := mErr.ErrorOrNil(); err != nil {
+		return diag.Errorf("error setting OBS bucket replication fields: %s", err)
+	}
+	return nil
+}
+
+func resourceObsBucketReplicationUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	obsClient, err := cfg.ObjectStorageClientWithSignature(region)
+	if err != nil {
+		return diag.Errorf("error creating OBS Client: %s", err)
+	}
+
+	updateOBSBucketReplicationChanges := []string{
+		"destination_bucket",
+		"agency",
+		"rule",
+	}
+	if d.HasChanges(updateOBSBucketReplicationChanges...) {
```

Comment:
region and bucket are ForceNew, so if the code goes in to Update func, that means at least one of "destination_bucket", "agency" and "rule" changed, we don't need to check again

Reply:
OK, I removed the Update function. update operation use the Create function